api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto

```diff
@@ -275,6 +275,31 @@ message ExternalProcessor {
   // backend stream lifetime. In this case, Envoy will eventually timeout the external processor stream according to this time limit.
   // The default value is 5000 milliseconds (5 seconds) if not specified.
   google.protobuf.Duration deferred_close_timeout = 19;
+
+  // [#not-implemented-hide:]
+  // Enables sending body to the side stream server before receives the header response from that server.
+  // It only works for STREAMED body processing mode. For any other body processing modes, it is ignored.
+  //
+  // Setting this flag to true is required in cases where the side stream server must wait for at least one
+  // chunk of body before sending the header response.
+  //
+  // The side stream server must not send a separate header response in this case. Instead, it encodes the
+  // response headers in the first body response and sends it back to Envoy.
```

Comment:
Thanks for reviewing! That's a good question.

My thoughts about this are:

1) The server should not send back a header response before seeing the body.  If they are allowed to do that, then there is no functional benefit to add this configuration.

2) If the server must wait for the body before it sends response, then it does not make sense for it to send a standalone header response first, then follow up with a body response. That wastes a RPC call. It is supported today that server can send a response contains both header and body mutation, as  in BUFFERED body mode: https://source.corp.google.com/piper///depot/google3/third_party/envoy/src/api/envoy/service/ext_proc/v3/external_processor.proto;rcl=660832859;l=289.

3) Also,  Envoy ext_proc state machine is already transitioned into wait-for-body-response state after it sends the body. In this state, it suppose just wait for body response, and a header response is deemed as "spurious message", which will shut down the external processing.  So, after the server sees both the header and body, if it is allowed to let the server sends the header response first, the, then Envoy state machine has to be changed to handle header response in wait-for-body-response state. It's doable, but probably not a good idea as it further complicates the state machine.

4) If we ask server must send one response which contains both header and body mutation, then it is well understandable that once Envoy receives such response, it can just go ahead forward it to the filter chain. Envoy can also deem a header only response is spurious, drops it and shut-down the external processing.    

5) Also, another open question is that, if we allow the server sends a header response first, follows up with body response, then after Envoy receives the header, should it buffer the header in Envoy and only send it to the filter chain after we see the body response? or send the header to the filter chain right away?

So, with all of the above thoughts, I think it is reasonable to add this limit, i.e, ask server sends both headers and first-chunk-body in one ProcessResponse, which simplify things a lot.

WDYT?

Reply:
I originally thought asking the SideStreamServer sends back a response includes both header and body is the best indication to know that the server had received the body before sending the response of headers, thus it is more secure.

However, after some discussion offline, I now agrees there is no need for Envoy to enforce security for the SideStreamServer.  The SideStreamServer should be allowed to send back the response whenever they think it is appropriate, and Envoy should be able to handle that. 

With that, so, the change will be using this newly added filter configuration to relax this check https://github.com/envoyproxy/envoy/blob/603bc61fbbf0ee968da5357bb258221609dc48ba/source/extensions/filters/http/ext_proc/ext_proc.cc#L492 when Envoy sees data.
And the SideStreamServer are allowed to send a separate header response, or encode the header response as header_mutation in the body response. Either way will be supported.  Once the responses comes back to Envoy, they will be forwarded to the filter chain right away.